runners/jet/src/main/java/org/apache/beam/runners/jet/DAGBuilder.java

```diff
@@ -0,0 +1,223 @@
+
+  DAG getDag() {
+    wireUp();
+    return dag;
+  }
+
+  void registerConstructionListeners(WiringListener listener) {
+    listeners.add(listener);
+  }
+
+  String newVertexId(String transformName) {
+    return vertexId++ + " (" + transformName + ")";
+  }
+
+  void registerCollectionOfEdge(String edgeId, String pCollId) {
+    String prevPCollId = pCollsOfEdges.put(edgeId, pCollId);
+    if (prevPCollId != null) {
+      throw new RuntimeException("Oops!");
+    }
+  }
+
+  void registerEdgeStartPoint(String edgeId, Vertex vertex, Coder coder) {
+    Objects.requireNonNull(edgeId);
+    Objects.requireNonNull(vertex);
+    Objects.requireNonNull(coder);
+
+    Vertex prevVertex = edgeStartPoints.put(edgeId, vertex);
+    if (prevVertex != null) {
+      throw new RuntimeException("Oops!");
+    }
+
+    Coder prevCoder = edgeCoders.put(edgeId, coder);
+    if (prevCoder != null) {
+      throw new RuntimeException("Oops!");
+    }
+  }
+
+  void registerEdgeEndPoint(String edgeId, Vertex vertex) {
+    edgeEndPoints.computeIfAbsent(edgeId, x -> new ArrayList<>()).add(vertex);
+  }
+
+  Vertex addVertex(String id, ProcessorMetaSupplier processorMetaSupplier) {
+    return dag.newVertex(id, processorMetaSupplier);
+  }
+
+  Vertex addVertex(String id, SupplierEx<Processor> processor) {
+    return dag.newVertex(id, processor)
+        .localParallelism(
+            localParallelism) // todo: quick and dirty hack for now, can't leave it like this
+    ;
+  }
+
+  private void wireUp() {
+    new WiringInstaller().wireUp();
+  }
+
+  /**
+   * Listener that can be registered with a {@link DAGBuilder} in order to be notified when edges
+   * are being registered.
+   */
+  public interface WiringListener {
+
+    void isOutboundEdgeOfVertex(Edge edge, String edgeId, String pCollId, String vertexId);
+
+    void isInboundEdgeOfVertex(Edge edge, String edgeId, String pCollId, String vertexId);
+  }
+
+  private class WiringInstaller {
+
+    private final Map<Vertex, Integer> inboundOrdinals = new HashMap<>();
+    private final Map<Vertex, Integer> outboundOrdinals = new HashMap<>();
+
+    void wireUp() {
+      Collection<String> edgeIds = new HashSet<>();
+      edgeIds.addAll(edgeStartPoints.keySet());
+      edgeIds.addAll(edgeEndPoints.keySet());
+
+      for (String edgeId : edgeIds) {
+        String pCollId = pCollsOfEdges.get(edgeId);
+        if (pCollId == null) {
+          throw new RuntimeException("Oops!");
+        }
+
+        Vertex sourceVertex = edgeStartPoints.get(edgeId);
+        if (sourceVertex == null) {
+          throw new RuntimeException("Oops!");
+        }
+
+        Coder edgeCoder = edgeCoders.get(edgeId);
+        if (edgeCoder == null) {
+          throw new RuntimeException("Oops!");
+        }
+
+        List<Vertex> destinationVertices =
+            edgeEndPoints.getOrDefault(edgeId, Collections.emptyList());
+        boolean sideInputEdge = edgeId.contains("PCollectionView"); // todo: this is a hack!
```

Comment:
Perhaps clean this up?

Reply:
Ok, that's fair. I will improve on it, but atm I have no idea how. I've taken a note and will add it to the next PR we're planning to submit.